cpp/src/Slice/MetadataValidator.h

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) ZeroC, Inc.
+
+#ifndef SLICE_UTIL_H
+#define SLICE_UTIL_H
+
+#include "Parser.h"
+
+#include <functional>
+
+namespace Slice
+{
+    /// TODO
+    using ValidationFunc = std::function<std::optional<std::string>(const SyntaxTreeBasePtr&, const MetadataPtr&)>;
+
+    /// TODO
+    class MetadataValidator final : public ParserVisitor
+    {
+    public:
+        MetadataValidator(std::string language, std::map<std::string, ValidationFunc> validators);
```

Comment:
I find this API very odd. This map of functions is like reimplementing dynamic dispatch "by hand".

A much simpler and idiomatic API is to make MetadataValidator an abstract base class with a pure virtual function such as:
```cpp
std::optional<string> validate(const MetadataPtr& metadata, const SyntaxTreeBasePtr& p) = 0;
```

And the derived class can use a map or case/switch statement to select the "validator" it wants to.

Reply:
I made the switch as you suggested. I guess I've gotten to used to avoiding class inheritance with Rust!

It is nice not to have this cumbersome `std::function` type, and gives implementations more flexibility.
But it does lead to more boilerplate in each of the compilers, since now they each need to:
1) Declare their own subclass of this type,
2) Implement their own 'switch' statement (or equivalent).

Both are very minor though.

If you have any other ideas on the API, let me know.
I want to get it nailed down before I spend any time rolling it out to all the languages.